### PR TITLE
Fix up incorrect data in etcd during Neutron resync

### PR DESCRIPTION
**Pull request by Lukasa**
This change is a pretty stupid solution to #571. Right now this is about half the work: it adds the function, but doesn't test it. I also want to do a refactor because a couple of these methods are massive now.

### CODE REVIEW

**Comment by Lukasa** [comment]
Ok, I believe this is now good to go. The refactor code is in #712, which should be considered only once this is merged.

I'm assigning this to @fasaxc for review, but it'd probably be good to do some live test of this before we merge. Let's not merge it until next week.

File: calico/openstack/mech_calico.py [hunk 1]
```diff
@@ -675,6 +683,52 @@ def resync_endpoints(self, context):
...
+                LOG.info("Failed to update deleted endpoint %s", endpoint.id)
+                continue
+
+            with context.session.begin(subtransactions=True):
+                port = self.db.get_port(context, endpoint.id)
+
+            # Get the data for both.
+            try:
+                etcd_data = json.loads(endpoint.data)
+            except Exception:
```
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: Would `ValueError` and `TypeError` be enough here?  I _think_ that's all loads can raise.

File: calico/openstack/mech_calico.py [hunk 2]
```diff
@@ -675,6 +683,52 @@ def resync_endpoints(self, context):
...
+                # If the JSON data is bad, just ignore it. We can't blow up
+                # here because that will trigger a new resync on another node
+                # that will blow *it* up as well. Just tolerate it.
+                LOG.exception("Bad JSON data in key %s", endpoint.key)
+                continue
```
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: `continue` doesn't seem right.  If the JSON data is bad shouldn't we be fixing it?

File: calico/openstack/mech_calico.py [hunk 3]
```diff
@@ -675,6 +683,52 @@ def resync_endpoints(self, context):
...
+
+            self.add_port_gateways(port, context)
```
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: Ignore if already covered in your refactor... These few lines feel duplicative.  Seems like there should be a neutron_port_to_etcd_endpoint function

Comment (Lukasa) [inline calico/openstack/mech_calico.py]: Already covered in the refactor. =)

File: calico/openstack/mech_calico.py [hunk 4]
```diff
@@ -730,6 +786,61 @@ def resync_profiles(self, context):
                 # means the profile was created or updated elsewhere.
                 continue
 
+        # Finally, reconcile the security profiles. This involves looping over
+        # them, grabbing their data, and then comparing that to what Neutron
+        # has.
+        changed_profiles = (p for p in profiles if p.id in reconcile_groups)
```
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: How about `profiles_to_reconcile`?  `changed_profiles` seems a bit off

File: calico/openstack/mech_calico.py [hunk 5]
```diff
@@ -730,6 +786,61 @@ def resync_profiles(self, context):
...
+            # Convert the etcd data into in-memory data structures.
+            try:
+                etcd_rules = json.loads(etcd_profile.rules_data)
+                etcd_tags = json.loads(etcd_profile.tags_data)
+            except Exception:
+                # If the JSON data is bad, just ignore it. We can't blow up
+                # here because that will trigger a new resync on another node
+                # that will blow *it* up as well. Just tolerate it.
+                LOG.exception("Bad JSON data in key %s", etcd_profile.key)
+                continue
```
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: Ditto, shouldn't we fix?

**Comment by fasaxc** [comment]
Few comments but didn't spot anything major.  Do you think we're stretching those named tuples a bit?  They're getting a bit unwieldy now they (sometimes) carry the data too.

File: calico/openstack/mech_calico.py (hunk 1 above)
Comment (Lukasa) [inline calico/openstack/mech_calico.py]: Let's assume the answer is yes. =D

File: calico/openstack/mech_calico.py (hunk 2 above)
Comment (Lukasa) [inline calico/openstack/mech_calico.py]: Yeah, we should. We do it in the refactor, weirdly, so I'm inclined to leave it in place for now and add it in the refactor code.

**Comment by Lukasa** [comment]
Ok @fasaxc, markups made. =)

File: calico/openstack/mech_calico.py (hunk 2 above)
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: OK

File: calico/openstack/t_etcd.py [hunk 6]
```diff
@@ -235,6 +256,27 @@ def provide_felix_config(self):
...
+        """
+        Get data for an endpoint out of etcd. This should be used on endpoints
+        returned from functions like ``get_endpoints``.
+
+        :param endpoint: An ``Endpoint`` class.
+        :return: A ``Endpoint`` class with ``data`` not None.
+        """
+        LOG.debug("Getting endpoint %s", endpoint.id)
+
+        result = self.client.read(endpoint.key, timeout=5)
```
Comment (fasaxc) [inline calico/openstack/t_etcd.py]: We're using these timeouts in a few places, worth wrapping up so we don't forget them and only have one place to change  at 3am when the timeout is too short.

**Comment by fasaxc** [comment]
Markups look good, one thought that just occurred to me but I don't need to see it again.

**Comment by fasaxc** [comment]
Good to merge after the 0.27 release, I think.

File: calico/openstack/t_etcd.py (hunk 6 above)
Comment (Lukasa) [inline calico/openstack/t_etcd.py]: Yeah, good idea.

**Comment by Lukasa** [comment]
And done. Will merge once we release.

**Pull request merged**